File: src/find_gcd_1979.rs
// https://leetcode.com/problems/find-greatest-common-divisor-of-array


pub fn find_gcd(nums: Vec<i32>) -> i32 {
    use itertools::MinMaxResult::MinMax;
    use itertools::Itertools;
    let MinMax(&min, &max) = nums.iter().minmax() else {
        return 0;
    };
    gcd(min, max)
}


fn gcd(mut a: i32, mut b: i32) -> i32 {
    if a == b {
        return a;
    }
    if b > a {
        std::mem::swap(&mut a, &mut b);
    }
    while b > 0 {
        let temp = a;
        a = b;
        b = temp % b;
    }
    a
}

#[cfg(test)]
mod tests {
    use super::find_gcd;

    #[test]
    fn official1() {
        let nums = vec![2, 5, 6, 9, 10];
        assert_eq!(2, find_gcd(nums));
    }

    #[test]
    fn official2() {
        let nums = vec![7, 5, 6, 8, 3];
        assert_eq!(1, find_gcd(nums));
    }

    #[test]
    fn official3() {
        let nums = vec![3, 3];
        assert_eq!(3, find_gcd(nums));
    }
}

File: cpp/src/find_gcd_1979.cpp
// https://leetcode.com/problems/find-greatest-common-divisor-of-array

#include "test.hpp"

#include <algorithm>
#include <numeric>
#include <vector>

    int findGcd(std::vector<int> nums) {
        const auto [min, max] = std::ranges::minmax(nums);
        return std::gcd(min, max);
    }

TEST(official1) {
    check_eq(2, findGcd({2, 5, 6, 9, 10}));
}

TEST(official2) {
    check_eq(1, findGcd({7, 5, 6, 8, 3}));
}

TEST(official3) {
    check_eq(3, findGcd({3, 3}));
}
